source/parser/x86_64/classifiers.hpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+// Spack Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+#pragma once
+
+#include <utility>
+
+#include "Type.h"
+#include "register_class.hpp"
+
+namespace smeagle::x86_64 {
+
+  struct Class {
+    RegisterClass lo, hi;
+    int pointer_indirections;
+    std::string name;
+  };
+
+  namespace st = Dyninst::SymtabAPI;
+
+  inline Class classify(st::typeScalar *t, int ptr_cnt) {
+    // size in BITS
+    const auto size = t->getSize() * 8;
+
+    if (ptr_cnt > 0) {
+      /*
+       * A pointer to a type X is converted to three abi_typelocation rules:
+       * 	1) 	A Pointer64 type at the base location and the base direction.
+       *
+       * 	2) 	A recursive conversion of application type X, at location “(base location)”
+       * 		(the base location wrapped in parentheses) and with direction ‘Export’.
+       *
+       * 	3)	A recursive conversion of application type X, at location “(base location)”
+       * 	(the base location wrapped in parentheses) and with direction ‘Import’.
+       *
+       * 	For example, an application type int* at base location %rax and direction ‘Import’
+       * would convert to: abi_typelocation(..., Import, Pointer64, “%rax”). abi_typelocation(...,
+       * Export, Integer32, “(%rax)”). abi_typelocation(..., Import, Integer32, “(%rax)”).
+       *
+       */
+      // TODO Should we integrate the directionality calculation here?
+      return {RegisterClass::INTEGER, RegisterClass::NO_CLASS, ptr_cnt, "Pointer64"};
```

Comment:
Why is this the entire block to return this hard coded thing? (is the logic just missing where the text is now?)

Reply:
This is a reminder comment to revisit some architectural designs. As I mentioned in my email last week and Matt brought up in the meeting, I think that we're getting too close to being x86 specific. The ABI model doc also has no discussion of multiple indirections- e.g., pointer to pointer or reference to pointer. I _think_ the current description will compose, but I'm not sure. In general, I think that now would be a good time for us to discuss some more high-level architecting for Smeagle in terms of separating platform-specific information from the generated model facts. I think we're definitely on the right track, overall, but this text about pointers made me really sit down and think about it. If you have some time this week, we should definitely chat!